### PR TITLE
Provide add/subtract methods for arithmetic calculations on addresses

### PR DESCRIPTION
**Pull request by victorhahncastell**
Unit tests provided and passing :-)

### CODE REVIEW

**Comment by bluemonk** [comment]
That's an impressive PR Victor, thanks a lot.

It will take me some time to go through it so bear with me for the moment please. At a first sight, I like the concept but i find the UI a bit confusing, maybe we can think of something better.

**Comment by victorhahncastell** [comment]
Sure, feel free to just change it or drop me any suggestions :-)

**Comment by victorhahncastell** [comment]
Hi Marco,

are you still planning to take this feature upstream? Is there anything I can adjust to have this better match your code?

As I currently work on a project using this feature I would love to see this in the main branch.

Best,
  Victor

**Comment by lumean** [comment]
Hi, I'd also like to use this for a project. Currently i just monkeypatch the most needed methods into the Ipaddress classes, but having this upstream would be much more convenient.
br manuel

**Comment by francisluong** [comment]
Having a look.  @victorhahncastell: if you're still plugged in can you re-merge master and resolve conflicts?

File: test/ipaddress/ipv4_test.rb [hunk 1]
```diff
@@ -13,13 +13,13 @@ def setup
       "10.0.0.1/255.255.255.0" => ["10.0.0.1", 24]}
     
     @invalid_ipv4 = ["10.0.0.256",
-                     "10.0.0.0.0",
```
Comment (francisluong) [inline test/ipaddress/ipv4_test.rb]: why the indent level changes?

File: lib/ipaddress/ipv6.rb [hunk 2]
```diff
@@ -327,6 +430,18 @@ def reverse
...
+    
+    #
+    # Returns this address' host part in unsigned 128bits format
+    #
+    #   ip = IPAddress::IPv6.new("fc42:1337:0:5::7/64")
+    #
+    #   ip.host_u128
+    #     #=> 7
+    #
+    def hostpart_u128
```
Comment (francisluong) [inline lib/ipaddress/ipv6.rb]: please supply unit testing for this method

File: lib/ipaddress/ipv6.rb [hunk 3]
```diff
@@ -315,8 +315,111 @@ def reverse
...
+    #   ip = IPAddress::IPv6.new("fc42:1337::/64")
+    #   ip.add(5).to_string
+    #     #=> "fc42:1337::5/64"
+    def add(oth, validating=true)
+      oth = oth.to_i if oth.kind_of? IPAddress::IPv6 # oth shall be integer
+      
+      new_obj = self.class.parse_u128(self.to_i + oth, prefix)
+      
+      if validating and self.network_u128 != new_obj.network_u128
+        raise ArgumentError, "Subnet (/#{@prefix}) is not large enough."
```
Comment (francisluong) [inline lib/ipaddress/ipv6.rb]: what do you think about changing this to a ```RuntimeError```.  The reason why I suggest this is that there isn't a problem with argument per se, rather you have reached a state of exhaustion.

File: lib/ipaddress/ipv6.rb [hunk 4]
```diff
@@ -315,8 +315,111 @@ def reverse
...
+      end
+      
+      return new_obj
```
Comment (francisluong) [inline lib/ipaddress/ipv6.rb]: https://github.com/bbatsov/ruby-style-guide#no-explicit-return

File: lib/ipaddress/ipv6.rb (hunk 2 above)
Comment (francisluong) [inline lib/ipaddress/ipv6.rb]: disregard... implied in testing of ```#hostpart```

File: test/ipaddress/ipv4_test.rb [hunk 5]
```diff
@@ -433,6 +433,57 @@ def test_method_supernet
     assert_equal "172.16.8.0/22", @ip.supernet(22).to_string
   end
 
+  def test_method_add
+    ip = IPAddress::IPv4.new("172.16.10.1/24")
+    assert_equal ip.add(5), IPAddress::IPv4.new("172.16.10.6/24")
```
Comment (francisluong) [inline test/ipaddress/ipv4_test.rb]: Please add parens, per https://github.com/bbatsov/ruby-style-guide#no-dsl-parens

Comment (lumean) [inline test/ipaddress/ipv4_test.rb]: There are no parens in the existing tests, should they be added everywhere? I'm currently integrating your suggestions.

**Comment by lumean** [comment]
Hi @francisluong , I would very much appreciate if we can continue this in #79 and then close them both. I hope this is ok with @victorhahncastell  since he didn't give any feedback so far.

**Comment by victorhahncastell** [comment]
@lumean: Sure, thanks for handling this :)

**Pull request closed without merging**